tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,594 @@
+        d3.select(this.$.tpgraph)
+          .append('svg')
+          .attr('width', SVG_WIDTH)
+          .attr('height', hostYDim * this._hostGridHeight
+                            + SVG_MARGIN.bottom + SVG_MARGIN.top)
+          .append('g')
+          .attr('transform',
+                'translate(' + SVG_MARGIN.left + ',' + SVG_MARGIN.top + ')');
+    const hostData = this.createHostData(hostXDim, hostYDim);
+    this.drawHostCards(
+        svg, hostData, this._hostGridWidth, this._hostGridHeight);
+    this.drawNodeCards(svg, data, colorScale);
+
+    // Creates separate groups, so that the z-index remains in the right order.
+    this._gLink = svg.append('svg:g').classed('link', true);
+
+    // Add a svg:defs for the arrow head.
+    svg.append('svg:defs')
+       .append('svg:marker')
+       .attr('id', 'arrow')
+       .attr('viewBox', '0 -5 10 10')
+       .attr('markerWidth', 5)
+       .attr('markerHeight', 5)
+       .attr('orient', 'auto')
+       .append('svg:path')
+       .style('stroke', 'red')
+       .style('fill', 'red')
+       .attr('d', 'M0,-5L10,0L0,5');
+    this.drawLabels(svg, chipXDims, chipYDims);
+    const legendYLoc =
+        this._hostGridHeight * Math.ceil(this._yDimension / HOST_Y_STRIDE) +
+        HOST_TO_HOST_MARGIN;
+    this.drawLegend(svg, legendYLoc, colorScale);
+  },
+  /**
+   * Returns the size of host grid, including the host card size and the margin
+   * between two hosts.
+   */
+  getHostGridSize(stride: number): number {
+    return HOST_TO_CHIP_MARGIN * 2 + CHIP_TO_CHIP_MARGIN * (stride - 1) +
+        CHIP_GRID_SIZE * stride + HOST_TO_HOST_MARGIN;
+  },
+  /**
+   * Returns the x-axis location for the xChip'th chip of the xHost'th host.
+   */
+  getChipXLoc: function(xHost: number, xChip: number): number {
+    return xHost * this._hostGridWidth + HOST_TO_CHIP_MARGIN +
+        xChip * (CHIP_GRID_SIZE + CHIP_TO_CHIP_MARGIN);
+  },
+  /**
+   * Returns the y-axis location for the yChip'th chip of the yHost'th host.
+   */
+  getChipYLoc: function(yHost: number, yChip: number): number {
+    return yHost * this._hostGridHeight + HOST_TO_CHIP_MARGIN +
+        yChip * (CHIP_GRID_SIZE + CHIP_TO_CHIP_MARGIN);
+  },
+  /**
+   * Returns the x-axis location for the xNode'th node of the xChip'th chip of
+   * the xHost'th host.
+   */
+  getNodeXLoc: function(xHost: number, xChip: number, xNode: number): number {
+    return this.getChipXLoc(xHost, xChip) + xNode * this._nodeGridWidth;
+  },
+  /**
+   * Returns the location for each host in the system.
+   */
+  createHostData: function(
+      hostXDim: number, hostYDim: number): Array<HostData> {
+    let hostData = [];
+    for (let i = 0; i < hostXDim; i++) {
+      for (let j = 0; j < hostYDim; j++) {
+        hostData.push({xdim: i, ydim: j});
+      }
+    }
+    return hostData;
+  },
+  /**
+   * Draw the labels on x-axis and y-axis.
+   */
+  drawLabels: function(svg: any, xdims: number[], ydims: number[]) {
+    let parent = this;
```

Comment:
because you use arrow function, this is unnecessary (and somewhat confusing; i.e., `this` is confusing -- what is it supposed to be? this component or will it be used inside another d3 that makes heavy use of `func.call(randomThis)`)

Reply:
Done. Yes, it is now unnecessary. I think originally I didn't use arrow function, so that need to refer to parent to access the member function in this polymer components. Now I can remove them :)